prac_07/project_management.py

```diff
@@ -0,0 +1,152 @@
+            filter_by_date(projects)
+        elif menu_selection == "A":
+            print("Let's add a new project")
+            new_project = get_new_project()
+            projects.append(new_project)
+        elif menu_selection == "U":
+            projects = update_project(projects)
+            print(projects)
+        else:
+            print("Invalid selection.")
+        print(MENU_TEXT)
+        menu_selection = input(">>> ").upper()
+    write_to_file(projects, DEFAULT_FILENAME)
+    print("Thank you for using custom-built project management software.")
+
+
+def display_projects(projects):
+    """Display projects grouped based on completion percentage."""
+    projects.sort()
+    print("Incomplete projects:")
+    for project in projects:
+        if project.completion_percentage < 100.0:
+            print(project)
+    print("Completed projects:")
+    for project in projects:
+        if project.completion_percentage == 100.0:
+            print(project)
+
+
+def process_file(filename, name_index, start_date_index, priority_index, cost_estimate_index,
+                 completion_percentage_index):
+    """Process data from file to store in a list of objects."""
+    projects = []
+    with open(filename, "r") as in_file:
+        in_file.readline()  # ignore headings
+        for line in in_file:
+            parts = line.strip().split("\t")
+            project = Project(parts[name_index], parts[start_date_index], int(parts[priority_index]),
+                              float(parts[cost_estimate_index]), float(parts[completion_percentage_index]))
+            projects.append(project)
+    return projects
+
+
+def write_to_file(projects, filename):
+    """Save data to a file."""
+    with open(filename, "w") as out_file:
+        print("Name\tStart Date\tPriority\tCost Estimate\tCompletion Percentage", file=out_file)
+        for project in projects:
+            print(f"{project.name}\t{project.start_date}\t{project.priority}\t{project.cost_estimate}\t"
+                  f"{project.completion_percentage}", file=out_file)
+
+
+def filter_by_date(projects):
+    """Filter list of projects when given a date."""
+    filtered_projects = []
+    user_date_string = input("Show projects that start after date (dd/mm/yy): ")
+    is_valid_date = False
+    while not is_valid_date:
+        try:
+            user_date = datetime.datetime.strptime(user_date_string, "%d/%m/%Y").date()
+            is_valid_date = True
+        except ValueError:
+            print("Invalid date.")
+            user_date_string = input("Show projects that start after date (dd/mm/yy): ")
+
+    for project in projects:
+        formatted_start_date = datetime.datetime.strptime(project.start_date, "%d/%m/%Y").date()
+        if formatted_start_date >= user_date:
+            filtered_projects.append(project)
+    filtered_projects.sort()
+    for filtered_project in filtered_projects:
+        print(filtered_project)
```

Comment:
The filtering works but the projects are not sorted by date, `filtered_projects.sort()` sorts by priority based on the `__lt__` in the project class.